gnovm/pkg/gnolang/type_check.go

```diff
@@ -215,6 +215,115 @@ func assertAssignableTo(xt, dt Type, autoNative bool) {
 	}
 }
 
+func checkConstantExpr(store Store, last BlockNode, vx Expr) {
+Main:
+	switch vx := vx.(type) {
+	case *NameExpr:
+		t := evalStaticTypeOf(store, last, vx)
+		if _, ok := t.(*ArrayType); ok {
+			break Main
+		}
+		panic(fmt.Sprintf("%s (variable of type %s) is not constant", vx.Name, t))
+	case *TypeAssertExpr:
+		panic(fmt.Sprintf("%s (comma, ok expression of type %s) is not constant", vx.String(), vx.Type))
+	case *IndexExpr:
+		panic(fmt.Sprintf("%s (variable of type %s) is not constant", vx.String(), vx.X))
+	case *CallExpr:
+		ift := evalStaticTypeOf(store, last, vx.Func)
+		switch baseOf(ift).(type) {
+		case *FuncType:
+			tup := evalStaticTypeOfRaw(store, last, vx).(*tupleType)
+
+			// check for built-in functions
+			if cx, ok := vx.Func.(*ConstExpr); ok {
+				if fv, ok := cx.V.(*FuncValue); ok {
+					if fv.PkgPath == uversePkgPath {
+						// TODO: should support min, max
+						switch {
+						case fv.Name == "len":
+							checkConstantExpr(store, last, vx.Args[0])
+							break Main
+						case fv.Name == "cap":
+							checkConstantExpr(store, last, vx.Args[0])
+							break Main
+						}
+					}
+				}
+			}
+
+			switch {
+			case len(tup.Elts) == 0:
+				panic(fmt.Sprintf("%s (no value) used as value", vx.String()))
+			case len(tup.Elts) == 1:
+				panic(fmt.Sprintf("%s (value of type %s) is not constant", vx.String(), tup.Elts[0]))
+			default:
+				panic(fmt.Sprintf("multiple-value %s (value of type %s) in single-value context", vx.String(), tup.Elts))
+			}
+		case *TypeType:
+			for _, arg := range vx.Args {
+				checkConstantExpr(store, last, arg)
+			}
+		case *NativeType:
+			panic("NativeType\n")
+		default:
+			panic(fmt.Sprintf(
+				"unexpected func type %v (%v)",
+				ift, reflect.TypeOf(ift)))
+		}
+	case *BinaryExpr:
+		checkConstantExpr(store, last, vx.Left)
+		checkConstantExpr(store, last, vx.Right)
```

Comment:
this one return an error in my PR, I think. Do you want I added as a test ?

Reply:
please see this: https://go.dev/play/p/dhxEqNA7p60

Reply:
I believe that when you run it, you'll encounter an error like `x (variable of type int) is not constant`. While it's not the same error, I think the description of the error in my PR better clarifies the issue we have.

Reply:
IMO, it would be more natural that the RHS is not const, rather than its child node.

Reply:
I find the current error message a bit unclear. For example, consider the following code:

```go
package main

const x = 1

var y = 1

const b = x == y

func main() {
	println("ok")
}
```

In Go, this will produce the error: `x == y (untyped bool value) is not constant`. However, in the PR, the error we get is: `y (variable of type int) is not constant`. 

This difference suggests that we should change our code to make it work, like so:

```go
package main

const x = 1

const y = 1

const b = x == y

func main() {
	println("ok")
}
```

The Go error message is more informative, indicating that the expression itself is not considered constant rather than simply pointing out that `y` is a variable.

Reply:
But I can the logic of the code if you think it is a blocker to be more go like